torchao/prototype/quantization/safetensors_support.py

```diff
@@ -0,0 +1,106 @@
+import json
+from typing import Dict
+
+import torch
+from safetensors.torch import load_file, save_file
+
+from torchao.prototype.quantization.safetensors_serialization import (
+    Float8TensorAttributeJSONEncoder,
+    object_from_dict,
+)
+from torchao.quantization import Float8Tensor
+
+
+def load_tensor_subclass_dict(file_path: str, device: str):
+    """
+    Load a dictionary of tensor subclasses from a safetensors file.
+
+    Args:
+        file_path: Path to the safetensors file
+
+    Returns:
+        Dictionary of reconstructed tensor subclasses
+    """
+    loaded_tensors = load_file(file_path, device)
+
+    with open(file_path, "rb") as f:
+        import struct
+
+        header_size = struct.unpack("<Q", f.read(8))[0]
+        header_bytes = f.read(header_size)
+        header = json.loads(header_bytes)
+        metadata = header.get("__metadata__", {})
+
+    if "tensor_names" not in metadata:
+        raise ValueError("No tensors found")
+
+    tensor_names = json.loads(metadata["tensor_names"])
+    result = {}
+
+    for tensor_name in tensor_names:
+        tensor_tensors = {}
+        for key, value in loaded_tensors.items():
+            if key.startswith(f"{tensor_name}:"):
+                # Remove the prefix
+                tensor_tensors[key[len(tensor_name) + 1 :]] = value
+
+        tensor_metadata = json.loads(metadata.get(tensor_name))
+        tensor_type = tensor_metadata.get("_type")
+
+        if tensor_type == Float8Tensor.__name__:
+            tensor_metadata["_data"].update(tensor_tensors)
+            result[tensor_name] = object_from_dict(tensor_metadata)
+        elif tensor_type == torch.Tensor.__name__:
+            result[tensor_name] = tensor_tensors["_data"]
+        else:
+            raise ValueError(f"Unsupported tensor type: {tensor_type}")
+
+    print(
+        f"Loaded {len(tensor_names)} tensor subclasses from {file_path} with metadata"
+    )
+    return result
+
+
+def save_tensor_subclass_dict(
+    tensor_dict: Dict[str, Dict[str, torch.Tensor]],
+    file_path: str,
+):
+    """
+    Save a dictionary of tensor subclasses with appropriate metadata.
```

Comment:
might be helpful to write down the end result, what is saved in file.

also for `load_tensor_subclass_dict`, it would be good to write down what is loaded as well I think